File: ios/Classes/convertor/InputConverter.swift
import PassioNutritionAISDK

struct InputConverter {
    
    func iconSizeFromString(_ iconString: String?) -> IconSize {
        guard let iconString = iconString,
              let iconSize = IconSize(rawValue: iconString) else {
            return .px90
        }
        return iconSize
    }
    
    func entityTypeFromString(_ typeString: String) -> PassioIDEntityType {
        guard let entityType = PassioIDEntityType(rawValue: typeString) else {
            return .item
        }
        return entityType
    }
    
    func mapToFoodDetectionConfigurationWithArgs(args: [String: Any]) -> FoodDetectionConfiguration {
        if let map = args["args"] as? [String: Any]  {
            return mapToFoodDetectionConfiguration(map: map)
        } else {
            return FoodDetectionConfiguration()
        }
    }
    
    func mapToVolumeDetectionMode(fromSting: String) -> VolumeDetectionMode {
        switch fromSting {
        case "auto":
            return .auto
        case "builtInDualWideCamera":
            return .dualWideCamera
        case "none":
            return .none
        default:
            return .none
        }
    }
    
    func mapToFoodDetectionConfiguration(map: [String: Any]) -> FoodDetectionConfiguration {
        
        func mapToFPS(fromString: String) -> PassioNutritionAI.FramesPerSecond {
            switch fromString {
            case "one":
                return .one
            case "two":
                return .two
            case "three", "four", "max":
                return .three
            default:
                return .two
            }
        }
        
        
        var configuration = FoodDetectionConfiguration()
        if let detectVisual = map["detectVisual"] as? Int {
            configuration.detectVisual = detectVisual  == 1 ? true : false
        }
        if let detectBarcodes = map["detectBarcodes"] as? Int {
            configuration.detectBarcodes = detectBarcodes == 1 ? true : false
        }
        if let detectPackagedFood = map["detectPackagedFood"] as? Int {
            configuration.detectPackagedFood = detectPackagedFood == 1 ? true : false
        }
        if let framesPerSecondString = map["framesPerSecond"] as? String {
            configuration.framesPerSecond = mapToFPS(fromString: framesPerSecondString)
        }
        if let volumeString = map ["volumeDetectionMode"] as? String {
            configuration.volumeDetectionMode = mapToVolumeDetectionMode(fromSting: volumeString)
        }
        return configuration
    }
    
    func mapToBoundingBox(box :[String: Any]) -> CGRect? {
        guard let x = box["left"] as? Double,
              let y = box["top"] as? Double,
              let width = box["width"] as? Double,
              let height = box["height"] as? Double else {
            return nil
        }
        return CGRect(x: x, y: y, width: width, height: height)
    }
    
    func mapToFetchFoodItemForDataInfo(map: [String: Any?]) -> (foodDataInfo: PassioFoodDataInfo?, servingQuantity: Double?, servingUnit: String?) {
        if let foodDataInfo = map["foodDataInfo"] as? [String: Any] {
            let passioFoodDataInfo = mapToPassioFoodDataInfo(map: foodDataInfo)
            let servingQuantity = map["servingQuantity"] as? Double
            let servingUnit = map["servingUnit"] as? String
            return (passioFoodDataInfo, servingQuantity, servingUnit)
        } else {
            return (nil, nil, nil)
        }
    }
    
    func mapToPassioFoodDataInfo(map: [String: Any?]) -> PassioFoodDataInfo? {
        guard let type = map["type"] as? String,
              let foodName = map["foodName"] as? String,
              let score = map["score"] as? Double,
              let brandName = map["brandName"] as? String,
              let iconId = map["iconId"] as? PassioID,
              let labelId = map["labelId"] as? String,
              let scoredName = map["scoredName"] as? String,
              let resultId = map["resultId"] as? String,
              let isShortName = map["isShortName"] as? Bool,
              let nutritionPreview = mapToPassioSearchNutritionPreview(map: map["nutritionPreview"] as? [String : Any]) else {
            return nil;
        }
        let tags = map["tags"] as? [String]
        return PassioFoodDataInfo(foodName: foodName, brandName: brandName, iconID: iconId, score: score, scoredName: scoredName, labelId: labelId, type: type, resultId: resultId, nutritionPreview: nutritionPreview, isShortName: isShortName, tags: tags)
    }
    
    private func mapToPassioSearchNutritionPreview(map :[String: Any?]?) -> PassioSearchNutritionPreview? {
        guard let map = map,
              let calories = map["calories"] as? Int,
              let carbs = map["carbs"] as? Double,
              let fat = map["fat"] as? Double,
              let fiber = map["fiber"] as? Double,
              let protein = map["protein"] as? Double,
              let servingUnit = map["servingUnit"] as? String,
              let servingQuantity = map["servingQuantity"] as? Double,
              let weightQuantity = map["weightQuantity"] as? Double,
              let weightUnit = map["weightUnit"] as? String else {
            return nil
        }
        return PassioSearchNutritionPreview(calories: calories, carbs: carbs, fat: fat, protein: protein, fiber: fiber, servingUnit: servingUnit, servingQuantity: servingQuantity, weightUnit: weightUnit, weightQuantity: weightQuantity)
    }
    
    func mapToPassioFoodItem(map :[String: Any?]?) -> PassioFoodItem? {
        guard let map = map else {
            return nil
        }
        let id = map["id"] as! String
        let scannedId = map["scannedId"] as! PassioNutritionAISDK.PassioID
        let name = map["name"] as! String
        let details = map["details"] as! String
        let iconId = map["iconId"] as! String
        let licenseCopy = ""
        let amount = mapToPassioFoodAmount(map: map["amount"] as! [String : Any?])
        
        let ingredientsInMap = map["ingredients"] as! [[String: Any?]]
        let ingredients = ingredientsInMap.map { mapToPassioIngredient(map: $0) }
        
        let refCode = (map["refCode"] as? String) ?? ""
        
        let tags = map["tags"] as? [String]
        
        return PassioFoodItem(id: id, scannedId: scannedId, name: name, details: details, iconId: iconId, licenseCopy: licenseCopy, amount: amount, ingredients: ingredients, refCode: refCode, tags: tags)
    }
    
    func mapToPassioFoodAmount(map :[String: Any?]) -> PassioFoodAmount {
        let servingSizesMap = map["servingSizes"] as! [[String: Any?]]
        let servingSizes = servingSizesMap.map { mapToPassioServingSize(map: $0)}
        
        let servingUnitsMap = map["servingUnits"] as! [[String: Any?]]
        let servingUnits = servingUnitsMap.map { mapToPassioServingUnit(map: $0)}
        
        let selectedUnit = map["selectedUnit"] as! String
        let selectedQuantity = map["selectedQuantity"] as! Double
        
        var amount = PassioFoodAmount(servingSizes: servingSizes, servingUnits: servingUnits)
        amount.selectedUnit = selectedUnit
        amount.selectedQuantity = selectedQuantity
        
        return amount
    }
    
    func mapToPassioServingSize(map :[String: Any?]) -> PassioServingSize {
        let quantity = map["quantity"] as! Double
        let unitName = map["unitName"] as! String
        
        return PassioServingSize(quantity: quantity, unitName: unitName)
    }
    
    func mapToPassioServingUnit(map :[String: Any?]) -> PassioServingUnit {
        let unitName = map["unitName"] as! String
        let weightInMap = map["weight"] as! [String: Any?]
        let weight = mapToMeasurementUnitMass(map: weightInMap)!
        
        return PassioServingUnit(unitName: unitName, weight: weight)
    }
    
    func mapToMeasurementUnitMass(map :[String: Any?]?) -> Measurement<UnitMass>? {
        guard let map = map else {
            return nil
        }
        let value = map["value"] as! Double
        let unitInString = map["unit"] as! String
        let unit = mapToUnitMass(unit: unitInString)
        
        return Measurement(value: value, unit: unit)
    }
    
    func mapToMeasurementUnitEnergy(map :[String: Any?]?) -> Measurement<UnitEnergy>? {
        guard let map = map else {
            return nil
        }
        let value = map["value"] as! Double
        let unitInString = map["unit"] as! String
        let unit = mapToUnitEnergy(unit: unitInString)
        
        return Measurement(value: value, unit: unit)
    }
    
    func mapToMeasurementUnitIU(map :[String: Any?]?) -> Double? {
        guard let map = map else {
            return nil
        }
        let value = map["value"] as! Double
            
        return value
    }
    
    func mapToUnitMass(unit: String) -> UnitMass {
        switch unit {
        case "kilograms":
            return UnitMass.kilograms
        case "milligrams":
            return UnitMass.milligrams
        case "micrograms":
            return UnitMass.micrograms
        case "milliliter":
            return UnitMass.grams
        default:
            return UnitMass.grams
        }
    }
    
    func mapToUnitEnergy(unit: String) -> UnitEnergy {
        return UnitEnergy.kilocalories
    }
    
    func mapToPassioIngredient(map: [String: Any?]) -> PassioIngredient {
        let id = map["id"] as! String
        let name = map["name"] as! String
        let iconId = map["iconId"] as! String
        let amount = mapToPassioFoodAmount(map: map["amount"] as! [String : Any?])
        let referenceNutrients = mapToPassioNutrients(map: map["referenceNutrients"] as! [String : Any?])
        let metadata = mapToPassioFoodMetadata(map: map["metadata"] as! [String : Any?])
        let refCode = map["refCode"] as! String
        let tags = map["tags"] as? [String]
        
        return PassioIngredient(id: id, name: name, iconId: iconId, amount: amount, referenceNutrients: referenceNutrients, metadata: metadata, refCode: refCode, tags: tags)
    }
    
    func mapToPassioNutrients(map: [String: Any?]) -> PassioNutrients {
        let weight = mapToMeasurementUnitMass(map: map["weight"] as? [String : Any?])!
        let alcohol = mapToMeasurementUnitMass(map: map["alcohol"] as? [String : Any?])
        let calcium = mapToMeasurementUnitMass(map: map["calcium"] as? [String : Any?])
        let calories = mapToMeasurementUnitEnergy(map: map["calories"] as? [String : Any?])
        let carbs = mapToMeasurementUnitMass(map: map["carbs"] as? [String : Any?])
        let cholesterol = mapToMeasurementUnitMass(map: map["cholesterol"] as? [String : Any?])
        let chromium = mapToMeasurementUnitMass(map: map["chromium"] as? [String : Any?])
        let fat = mapToMeasurementUnitMass(map: map["fat"] as? [String : Any?])
        let fibers = mapToMeasurementUnitMass(map: map["fibers"] as? [String : Any?])
        let folicAcid = mapToMeasurementUnitMass(map: map["folicAcid"] as? [String : Any?])
        let iodine = mapToMeasurementUnitMass(map: map["iodine"] as? [String : Any?])
        let iron = mapToMeasurementUnitMass(map: map["iron"] as? [String : Any?])
        let magnesium = mapToMeasurementUnitMass(map: map["magnesium"] as? [String : Any?])
        let monounsaturatedFat = mapToMeasurementUnitMass(map: map["monounsaturatedFat"] as? [String : Any?])
        let phosphorus = mapToMeasurementUnitMass(map: map["phosphorus"] as? [String : Any?])
        let polyunsaturatedFat = mapToMeasurementUnitMass(map: map["polyunsaturatedFat"] as? [String : Any?])
        let potassium = mapToMeasurementUnitMass(map: map["potassium"] as? [String : Any?])
        let proteins = mapToMeasurementUnitMass(map: map["proteins"] as? [String : Any?])
        let satFat = mapToMeasurementUnitMass(map: map["satFat"] as? [String : Any?])
        let selenium = mapToMeasurementUnitMass(map: map["selenium"] as? [String : Any?])
        let sodium = mapToMeasurementUnitMass(map: map["sodium"] as? [String : Any?])
        let sugars = mapToMeasurementUnitMass(map: map["sugars"] as? [String : Any?])
        let sugarsAdded = mapToMeasurementUnitMass(map: map["sugarsAdded"] as? [String : Any?])
        let sugarAlcohol = mapToMeasurementUnitMass(map: map["sugarAlcohol"] as? [String : Any?])
        let transFat = mapToMeasurementUnitMass(map: map["transFat"] as? [String : Any?])
        let vitaminA = mapToMeasurementUnitIU(map: map["vitaminA"] as? [String : Any?])
        let vitaminB6 = mapToMeasurementUnitMass(map: map["vitaminB6"] as? [String : Any?])
        let vitaminB12 = mapToMeasurementUnitMass(map: map["vitaminB12"] as? [String : Any?])
        let vitaminB12Added = mapToMeasurementUnitMass(map: map["vitaminB12Added"] as? [String : Any?])
        let vitaminC = mapToMeasurementUnitMass(map: map["vitaminC"] as? [String : Any?])
        let vitaminD = mapToMeasurementUnitMass(map: map["vitaminD"] as? [String : Any?])
        let vitaminE = mapToMeasurementUnitMass(map: map["vitaminE"] as? [String : Any?])
        let vitaminEAdded = mapToMeasurementUnitMass(map: map["vitaminEAdded"] as? [String : Any?])
        let vitaminKDihydrophylloquinone = mapToMeasurementUnitMass(map: map["vitaminKDihydrophylloquinone"] as? [String : Any?])
        let vitaminKMenaquinone4 = mapToMeasurementUnitMass(map: map["vitaminKMenaquinone4"] as? [String : Any?])
        let vitaminKPhylloquinone = mapToMeasurementUnitMass(map: map["vitaminKPhylloquinone"] as? [String : Any?])
        let vitaminARAE = mapToMeasurementUnitMass(map: map["vitaminARAE"] as? [String : Any?])
        let zinc = mapToMeasurementUnitMass(map: map["zinc"] as? [String : Any?])
        
        return PassioNutrients(fat: fat, satFat: satFat, monounsaturatedFat: monounsaturatedFat, polyunsaturatedFat: polyunsaturatedFat, proteins: proteins, carbs: carbs, calories: calories, cholesterol: cholesterol, sodium: sodium, fibers: fibers, transFat: transFat, sugars: sugars, sugarsAdded: sugarsAdded, alcohol: alcohol, iron: iron, vitaminC: vitaminC, vitaminA: vitaminA, vitaminA_RAE: vitaminARAE, vitaminD: vitaminD, vitaminB6: vitaminB6, vitaminB12: vitaminB12, vitaminB12Added: vitaminB12Added, vitaminE: vitaminE, vitaminEAdded: vitaminEAdded, iodine: iodine, calcium: calcium, potassium: potassium, magnesium: magnesium, phosphorus: phosphorus, sugarAlcohol: sugarAlcohol, zinc: zinc, selenium: selenium, folicAcid: folicAcid, chromium: chromium, vitaminKPhylloquinone: vitaminKPhylloquinone, vitaminKMenaquinone4: vitaminKMenaquinone4, vitaminKDihydrophylloquinone: vitaminKDihydrophylloquinone, weight: weight)
    }
    
    func mapToPassioFoodMetadata(map: [String: Any?]) -> PassioFoodMetadata {
        let foodOriginsInMap = map["foodOrigins"] as? [[String: Any?]]
        let foodOrigins = foodOriginsInMap?.map { mapToPassioFoodOrigin(map: $0)! }
        
        let barcode = map["barcode"] as? Barcode
        let ingredientsDescription = map["ingredientsDescription"] as? String
        let tags = map["tags"] as? [String]
        
        return PassioFoodMetadata(foodOrigins: foodOrigins, barcode: barcode, ingredientsDescription: ingredientsDescription, tags: tags)
    }
    
    func mapToPassioFoodOrigin(map: [String: Any?]) -> PassioFoodOrigin? {
        let id = map["id"] as! String
        let source = map["source"] as! String
        let licenseCopy = map["licenseCopy"] as? String
        
        return PassioFoodOrigin(id: id, source: source, licenseCopy: licenseCopy)
    }
    
    
}

extension Dictionary {
    
    var toJSON: String? {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: self, options: .prettyPrinted)
            if let jsonString = String(data: jsonData, encoding: .utf8) {
                return jsonString
            }
        } catch {
            print("Unable to convert to json strin:- \(error.localizedDescription)")
        }
        return nil
    }

}

File: android/src/main/kotlin/ai/passio/nutrition_ai/converter/InputConverter.kt
package ai.passio.nutrition_ai.converter

import ai.passio.passiosdk.core.config.PassioConfiguration
import ai.passio.passiosdk.core.icons.IconSize
import ai.passio.passiosdk.passiofood.Barcode
import ai.passio.passiosdk.passiofood.FoodDetectionConfiguration
import ai.passio.passiosdk.passiofood.PassioFoodDataInfo
import ai.passio.passiosdk.passiofood.PassioID
import ai.passio.passiosdk.passiofood.PassioImageResolution
import ai.passio.passiosdk.passiofood.PassioSDK
import ai.passio.passiosdk.passiofood.PassioSearchNutritionPreview
import ai.passio.passiosdk.passiofood.data.measurement.Grams
import ai.passio.passiosdk.passiofood.data.measurement.KiloCalories
import ai.passio.passiosdk.passiofood.data.measurement.Kilograms
import ai.passio.passiosdk.passiofood.data.measurement.Micrograms
import ai.passio.passiosdk.passiofood.data.measurement.Milligrams
import ai.passio.passiosdk.passiofood.data.measurement.Milliliters
import ai.passio.passiosdk.passiofood.data.measurement.UnitEnergy
import ai.passio.passiosdk.passiofood.data.measurement.UnitMass
import ai.passio.passiosdk.passiofood.data.model.PassioAdvisorFoodInfo
import ai.passio.passiosdk.passiofood.data.model.PassioAdvisorResponse
import ai.passio.passiosdk.passiofood.data.model.PassioFoodAmount
import ai.passio.passiosdk.passiofood.data.model.PassioFoodItem
import ai.passio.passiosdk.passiofood.data.model.PassioFoodMetadata
import ai.passio.passiosdk.passiofood.data.model.PassioFoodOrigin
import ai.passio.passiosdk.passiofood.data.model.PassioIDEntityType
import ai.passio.passiosdk.passiofood.data.model.PassioIngredient
import ai.passio.passiosdk.passiofood.data.model.PassioNutrients
import ai.passio.passiosdk.passiofood.data.model.PassioServingSize
import ai.passio.passiosdk.passiofood.data.model.PassioServingUnit
import android.content.Context
import android.graphics.RectF
import android.net.Uri

@Suppress("UNCHECKED_CAST")
fun <T> Map<String, Any?>.mapContains(key: String, block: (value: T) -> Unit) {
    if (this.contains(key)) {
        block(this[key]!! as T)
    }
}

@Suppress("UNCHECKED_CAST")
fun <T> Map<String, Any?>.mapContainsOptional(key: String, block: (value: T?) -> Unit) {
    if (this.contains(key)) {
        block(this[key] as T?)
    }
}

fun mapToPassioConfiguration(context: Context, map: Map<String, Any?>): PassioConfiguration {
    return PassioConfiguration(context, map["key"] as String).apply {
        map.mapContains<Boolean>("sdkDownloadsModels") { this.sdkDownloadsModels = it }
        map.mapContains<Int>("debugMode") { this.debugMode = it }
        map.mapContains<Boolean>("allowInternetConnection") { this.allowInternetConnection = it }
        map.mapContains<Boolean>("remoteOnly") { this.remoteOnly = it }
        map.mapContainsOptional<List<String>>("filesLocalURLs") { files ->
            val uris = files?.map { Uri.parse(it) }
            this.localFiles = uris
        }
        map.mapContainsOptional<String?>("proxyUrl") { this.proxyUrl = it }
        map.mapContainsOptional<Map<String, String>?>("proxyHeaders") { this.proxyHeaders = it }
    }
}

fun mapToFoodDetectionConfiguration(map: Map<String, Any?>): FoodDetectionConfiguration {
    return FoodDetectionConfiguration().apply {
        map.mapContains<Boolean>("detectVisual") { this.detectVisual = it }
        map.mapContains<Boolean>("detectBarcodes") { this.detectBarcodes = it }
        map.mapContains<Boolean>("detectPackagedFood") { this.detectPackagedFood = it }
        map.mapContains<String>("framesPerSecond") {
            this.framesPerSecond = framesPerSecondFromString(it)
        }
    }
}

fun iconSizeFromString(iconString: String): IconSize {
    return when (iconString) {
        "px90" -> IconSize.PX90
        "px180" -> IconSize.PX180
        "px360" -> IconSize.PX360
        else -> throw java.lang.IllegalArgumentException("No known icon size: $iconString")
    }
}

fun entityTypeFromString(typeString: String): PassioIDEntityType {
    return when (typeString) {
        "group" -> PassioIDEntityType.group
        "item" -> PassioIDEntityType.item
        "recipe" -> PassioIDEntityType.recipe
        "barcode" -> PassioIDEntityType.barcode
        "packagedFoodCode" -> PassioIDEntityType.packagedFoodCode
        "nutritionFacts" -> PassioIDEntityType.nutritionFacts
        else -> throw java.lang.IllegalArgumentException("No known entity type: $typeString")
    }
}

fun framesPerSecondFromString(fps: String): PassioSDK.FramesPerSecond {
    return when (fps) {
        "one" -> PassioSDK.FramesPerSecond.ONE
        "two" -> PassioSDK.FramesPerSecond.TWO
        "three" -> PassioSDK.FramesPerSecond.MAX
        "four" -> PassioSDK.FramesPerSecond.MAX
        "max" -> PassioSDK.FramesPerSecond.MAX
        else -> throw java.lang.IllegalArgumentException("No known fps: $fps")
    }
}

fun mapToRectF(map: Map<String, Any?>): RectF {
    return RectF(
        (map["left"] as? Double)?.toFloat() ?: 0f,
        (map["top"] as? Double)?.toFloat() ?: 0f,
        ((map["left"] as? Double) ?: 0.0).toFloat() + ((map["width"] as? Double) ?: 0.0).toFloat(),
        ((map["top"] as? Double) ?: 0.0).toFloat() + ((map["height"] as? Double) ?: 0.0).toFloat(),
    )
}

fun mapToFetchFoodItemForDataInfo(map: Map<String, Any?>): Triple<PassioFoodDataInfo, Double?, String?> {
    val foodDataInfo = map["foodDataInfo"] as Map<String, Any>
    val passioFoodDataInfo = mapToPassioFoodDataInfo(foodDataInfo)
    val servingQuantity = map["servingQuantity"] as? Double
    val servingUnit = map["servingUnit"] as? String
    return Triple(passioFoodDataInfo, servingQuantity, servingUnit)
}

fun mapToPassioFoodDataInfo(map: Map<String, Any?>): PassioFoodDataInfo {
    return PassioFoodDataInfo(
        map["refCode"] as String,
        map["foodName"] as String,
        map["brandName"] as String,
        map["iconId"] as PassioID,
        map["score"] as Double,
        map["scoredName"] as String,
        map["labelId"] as String,
        map["type"] as String,
        map["resultId"] as String,
        map["isShortName"] as Boolean,
        mapToPassioSearchNutritionPreview(map["nutritionPreview"] as Map<String, Any?>),
        map["tags"] as List<String>?,
    )
}

private fun mapToPassioSearchNutritionPreview(map: Map<String, Any?>): PassioSearchNutritionPreview {
    return PassioSearchNutritionPreview(
        map["calories"] as Int,
        map["carbs"] as Double,
        map["protein"] as Double,
        map["fat"] as Double,
        map["fiber"] as Double,
        map["servingUnit"] as String,
        map["servingQuantity"] as Double,
        map["weightUnit"] as String,
        map["weightQuantity"] as Double,
    )
}

fun mapToPassioAdvisorResponse(map: Map<String, Any?>): PassioAdvisorResponse {
    var tools: List<String>? = null
    var extractedIngredients: List<PassioAdvisorFoodInfo>? = null

    map.mapContainsOptional<List<String>?>("tools") { tool ->
        tools = tool
    }
    map.mapContainsOptional<List<PassioAdvisorFoodInfo>?>("extractedIngredients") { ingredients ->
        extractedIngredients = ingredients
    }

    return PassioAdvisorResponse(
        map["threadId"] as String,
        map["messageId"] as String,
        map["markupContent"] as String,
        map["rawContent"] as String,
        tools,
        extractedIngredients
    )
}

fun passioImageResolutionFromString(resolutionString: String): PassioImageResolution {
    return when (resolutionString) {
        "res_512" -> PassioImageResolution.RES_512
        "res_1080" -> PassioImageResolution.RES_1080
        "full" -> PassioImageResolution.FULL
        else -> throw java.lang.IllegalArgumentException("No known PassioImageResolution: $resolutionString")
    }
}

fun mapToPassioFoodItem(map: Map<String, Any?>): PassioFoodItem {
    val id = map["id"] as String
    val refCode = map["refCode"] as String
    val name = map["name"] as String
    val details = map["details"] as String
    val iconId = map["iconId"] as String
    val amount = mapToPassioFoodAmount(map["amount"] as Map<String, Any?>)

    // Ingredients
    val ingredientsInMap = map["ingredients"] as List<Map<String, Any?>>
    val ingredients = ingredientsInMap.map { mapToPassioIngredient(it) }

    return PassioFoodItem(id, refCode, name, details, iconId, amount, ingredients)
}

fun mapToPassioIngredient(map: Map<String, Any?>): PassioIngredient {
    val id = map["id"] as String
    val refCode = map["refCode"] as String
    val name = map["name"] as String
    val iconId = map["iconId"] as String
    val amount = mapToPassioFoodAmount(map["amount"] as Map<String, Any?>)
    val referenceNutrients = mapToPassioNutrients(map["referenceNutrients"] as Map<String, Any?>)
    val metadata = mapToPassioFoodMetadata(map["metadata"] as Map<String, Any?>)

    return PassioIngredient(id, refCode, name, iconId, amount, referenceNutrients, metadata)
}

fun mapToPassioFoodAmount(map: Map<String, Any?>): PassioFoodAmount {
    // servingSizes
    val servingSizesInMap = map["servingSizes"] as List<Map<String, Any?>>
    val servingSizes = servingSizesInMap.map { mapToPassioServingSize(it) }

    // servingSizes
    val servingUnitsInMap = map["servingUnits"] as List<Map<String, Any?>>
    val servingUnits = servingUnitsInMap.map { mapToPassioServingUnit(it) }

    val selectedUnit = map["selectedUnit"] as String
    val selectedQuantity = map["selectedQuantity"] as Double

    val amount = PassioFoodAmount(servingSizes, servingUnits)
    amount.selectedUnit = selectedUnit
    amount.selectedQuantity = selectedQuantity

    return amount
}

fun mapToPassioServingSize(map: Map<String, Any?>): PassioServingSize {
    val quantity = map["quantity"] as Double
    val unitName = map["unitName"] as String
    return PassioServingSize(quantity, unitName)
}

fun mapToPassioServingUnit(map: Map<String, Any?>): PassioServingUnit {
    val unitName = map["unitName"] as String
    val weight = mapToUnitMass(map["weight"] as Map<String, Any?>)!!
    return PassioServingUnit(unitName, weight)
}

fun mapToUnitMass(map: Map<String, Any?>?): UnitMass? {
    if(map == null) return null
    val value = map["value"] as Double

    val unitInString = map["unit"] as String
    val unit = unitInString.unitFromString()

    return UnitMass(unit, value)
}

fun mapToUnitEnergy(map: Map<String, Any?>?): UnitEnergy? {
    if(map == null) return null
    val value = map["value"] as Double

    val unitInString = map["unit"] as String
    val unit = unitInString.unitFromString()

    return UnitEnergy(unit, value)
}

fun mapToUnitIU(map: Map<String, Any?>?): Double? {
    if(map == null) return null
    val value = map["value"] as Double

    return  value
}

private fun String.unitFromString(): ai.passio.passiosdk.passiofood.data.measurement.Unit {
    return when (this) {
        "kilograms" -> Kilograms
        "grams" -> Grams
        "milligrams" -> Milligrams
        "micrograms" -> Micrograms
        "milliliter" -> Milliliters
        "kilocalories" -> KiloCalories
        else -> throw java.lang.IllegalArgumentException("No known unit: $this")
    }
}

fun mapToPassioNutrients(map: Map<String, Any?>): PassioNutrients {
    val weight = mapToUnitMass(map["weight"] as Map<String, Any?>)!!
    val alcohol = mapToUnitMass(map["alcohol"] as? Map<String, Any?>)
    val calcium = mapToUnitMass(map["calcium"] as? Map<String, Any?>)
    val calories = mapToUnitEnergy(map["calories"] as? Map<String, Any?>)
    val carbs = mapToUnitMass(map["carbs"] as? Map<String, Any?>)
    val cholesterol = mapToUnitMass(map["cholesterol"] as? Map<String, Any?>)
    val chromium = mapToUnitMass(map["chromium"] as? Map<String, Any?>)
    val fat = mapToUnitMass(map["fat"] as? Map<String, Any?>)
    val fibers = mapToUnitMass(map["fibers"] as? Map<String, Any?>)
    val folicAcid = mapToUnitMass(map["folicAcid"] as? Map<String, Any?>)
    val iodine = mapToUnitMass(map["iodine"] as? Map<String, Any?>)
    val iron = mapToUnitMass(map["iron"] as? Map<String, Any?>)
    val magnesium = mapToUnitMass(map["magnesium"] as? Map<String, Any?>)
    val monounsaturatedFat = mapToUnitMass(map["monounsaturatedFat"] as? Map<String, Any?>)
    val phosphorus = mapToUnitMass(map["phosphorus"] as? Map<String, Any?>)
    val polyunsaturatedFat = mapToUnitMass(map["polyunsaturatedFat"] as? Map<String, Any?>)
    val potassium = mapToUnitMass(map["potassium"] as? Map<String, Any?>)
    val proteins = mapToUnitMass(map["proteins"] as? Map<String, Any?>)
    val satFat = mapToUnitMass(map["satFat"] as? Map<String, Any?>)
    val selenium = mapToUnitMass(map["selenium"] as? Map<String, Any?>)
    val sodium = mapToUnitMass(map["sodium"] as? Map<String, Any?>)
    val sugars = mapToUnitMass(map["sugars"] as? Map<String, Any?>)
    val sugarsAdded = mapToUnitMass(map["sugarsAdded"] as? Map<String, Any?>)
    val sugarAlcohol = mapToUnitMass(map["sugarAlcohol"] as? Map<String, Any?>)
    val transFat = mapToUnitMass(map["transFat"] as? Map<String, Any?>)
    val vitaminA = mapToUnitIU(map["vitaminA"] as? Map<String, Any?>)
    val vitaminB6 = mapToUnitMass(map["vitaminB6"] as? Map<String, Any?>)
    val vitaminB12 = mapToUnitMass(map["vitaminB12"] as? Map<String, Any?>)
    val vitaminB12Added = mapToUnitMass(map["vitaminB12Added"] as? Map<String, Any?>)
    val vitaminC = mapToUnitMass(map["vitaminC"] as? Map<String, Any?>)
    val vitaminD = mapToUnitMass(map["vitaminD"] as? Map<String, Any?>)
    val vitaminE = mapToUnitMass(map["vitaminE"] as? Map<String, Any?>)
    val vitaminEAdded = mapToUnitMass(map["vitaminEAdded"] as? Map<String, Any?>)
    val vitaminKDihydrophylloquinone =
        mapToUnitMass(map["vitaminKDihydrophylloquinone"] as? Map<String, Any?>)
    val vitaminKMenaquinone4 = mapToUnitMass(map["vitaminKMenaquinone4"] as? Map<String, Any?>)
    val vitaminKPhylloquinone = mapToUnitMass(map["vitaminKPhylloquinone"] as? Map<String, Any?>)
    val vitaminARAE = mapToUnitMass(map["vitaminARAE"] as? Map<String, Any?>)
    val zinc = mapToUnitMass(map["zinc"] as? Map<String, Any?>)

    return PassioNutrients(
        weight,
        fat,
        satFat,
        monounsaturatedFat,
        polyunsaturatedFat,
        proteins,
        carbs,
        calories,
        cholesterol,
        sodium,
        fibers,
        transFat,
        sugars,
        sugarsAdded,
        alcohol,
        iron,
        vitaminC,
        vitaminD,
        vitaminB6,
        vitaminB12,
        vitaminB12Added,
        vitaminE,
        vitaminEAdded,
        iodine,
        calcium,
        potassium,
        magnesium,
        phosphorus,
        sugarAlcohol,
        vitaminA,
        vitaminARAE,
        vitaminKPhylloquinone,
        vitaminKMenaquinone4,
        vitaminKDihydrophylloquinone,
        zinc,
        chromium,
        selenium,
        folicAcid,
    )
}

fun mapToPassioFoodMetadata(map: Map<String, Any?>): PassioFoodMetadata {
    val foodOriginsInMap = map["foodOrigins"] as List<Map<String, Any?>>?
    val foodOrigins = foodOriginsInMap?.map { mapToPassioFoodOrigin(it) }

    val barcode = map["barcode"] as Barcode?
    val ingredientsDescription = map["ingredientsDescription"] as String?
    val tags = map["ingredientsDescription"] as List<String>?

    return PassioFoodMetadata(foodOrigins, barcode, ingredientsDescription, tags)
}

fun mapToPassioFoodOrigin(map: Map<String, Any?>): PassioFoodOrigin {
    val id = map["id"] as String
    val source = map["source"] as String
    val licenseCopy = map["licenseCopy"] as String?

    return PassioFoodOrigin(id, source, licenseCopy)
}